components/script/dom/bindings/interface.rs

```diff
@@ -158,6 +169,82 @@ pub unsafe fn create_global_object(
     JS_FireOnNewGlobalObject(cx, rval.handle());
 }
 
+// https://html.spec.whatwg.org/multipage/#htmlconstructor
+pub unsafe fn html_constructor<T>(window: &Window, call_args: CallArgs) -> Fallible<Root<T>>
+                                  where T: DomObject + DerivedFrom<HTMLElement> {
+    let document = window.Document();
+
+    // Step 1
+    let registry = window.CustomElements();
+
+    // Step 2 in codegen
+
+    // Step 3
+    rooted!(in(window.get_cx()) let new_target = call_args.new_target().to_object());
+    let definition = match registry.lookup_definition_by_constructor(new_target.handle()) {
+        Some(definition) => definition,
+        None => return Err(Error::Type("No custom element definition found for new.target".to_owned())),
+    };
+
+    rooted!(in(window.get_cx()) let callee = UnwrapObject(call_args.callee(), 1));
+    if callee.is_null() {
+        return Err(Error::Security);
+    }
+
+    {
+        let _ac = JSAutoCompartment::new(window.get_cx(), callee.get());
+
+        if definition.is_autonomous() {
+            // Step 4
+            // Since this element is autonomous, its active function object must be the HTMLElement
+
+            // Retreive the constructor object for HTMLElement
+            rooted!(in(window.get_cx()) let mut constructor = ptr::null_mut());
+            rooted!(in(window.get_cx()) let global_object = CurrentGlobalOrNull(window.get_cx()));
+            HTMLElementBinding::GetConstructorObject(window.get_cx(), global_object.handle(), constructor.handle_mut());
+            if constructor.is_null() {
+                return Err(Error::JSFailed);
+            }
+
+            // Callee must be the same constructor object as HTMLElement
+            if constructor.get() != callee.get() {
+                return Err(Error::Type("Active function object is not HTMLElement".to_owned()));
+            }
+        } else {
+            // TODO: Step 5
```

Comment:
Yeah, not generating `GetConstructorObject` for leaf interfaces was just me trimming a bit of unneeded codegen, but if it's not unneeded anymore, just make it generate the function again.